File: apps/backend/src/entitlements.ts
import {
  PLANS,
  type FeatureKey,
  type PlanKey,
  featureLimit as planFeatureLimit,
  getPlan,
} from "@yomi/shared/plans"

type EntitlementUser = {
  id?: string | null
  email?: string | null
  role?: string | null
  plan?: string | null
}

type RenewalUser = EntitlementUser & {
  createdAt?: Date | null
  trialEndDate?: Date | null
  currentPeriodEnd?: Date | null
}

// "none" — a paid plan with no billing period on record yet.
export type CreditRenewal = {
  kind: "renewal" | "none"
  at: Date | null
}

const TRIAL_MS = 30 * 24 * 60 * 60 * 1000

export function effectiveRoleForUser(user: EntitlementUser): string {
  return user.role ?? "user"
}

export function effectivePlanForUser(user: EntitlementUser): string {
  const plan = user.plan ?? "explore"
  return plan in PLANS ? plan : "explore"
}

// Explore renews every month like a paid plan now (see explore-renewal.ts's cron
// sweep) rather than expiring after a one-off trial grant, so it reports the same
// "renewal" kind — trialEndDate is the recurring renewal date, not an end date.
// Paid plans are re-granted by the Dodo cycle webhook at currentPeriodEnd (billing.ts).
export function creditRenewal(user: RenewalUser): CreditRenewal {
  if (effectivePlanForUser(user) === "explore") {
    const trialEnd =
      user.trialEndDate ?? (user.createdAt ? new Date(user.createdAt.getTime() + TRIAL_MS) : null)
    return trialEnd ? { kind: "renewal", at: trialEnd } : { kind: "none", at: null }
  }

  return user.currentPeriodEnd
    ? { kind: "renewal", at: user.currentPeriodEnd }
    : { kind: "none", at: null }
}

export function requestLimitForUser(user: EntitlementUser): number | null {
  const plan = effectivePlanForUser(user) as PlanKey
  return planFeatureLimit(plan, "chat") as number
}

export function featureLimitForUser(user: EntitlementUser, feature: FeatureKey): number | null {
  const plan = effectivePlanForUser(user) as PlanKey
  return planFeatureLimit(plan, feature) as number
}

export function getPlanConfig(user: EntitlementUser) {
  const plan = effectivePlanForUser(user)
  return getPlan(plan)
}

export function hasBillablePlanAccess(
  user: EntitlementUser & {
    subscriptionStatus?: string | null
    currentPeriodEnd?: Date | null
    trialEndDate?: Date | null
  },
): boolean {
  const plan = effectivePlanForUser(user)

  if (plan === "explore") {
    console.warn(
      `[hasBillablePlanAccess] explore trial check: trialEndDate=${user.trialEndDate} now=${new Date()} result=${!user.trialEndDate ? "false(no trialEndDate)" : new Date() < user.trialEndDate}`,
    )
    if (!user.trialEndDate) return false
    return new Date() < user.trialEndDate
  }

  if (user.subscriptionStatus === "active" || user.subscriptionStatus === "trialing") return true

  if (user.subscriptionStatus === "past_due") {
    const refMs = user.currentPeriodEnd?.getTime() ?? Date.now()
    const graceEnd = new Date(refMs + 7 * 24 * 60 * 60 * 1000)
    if (new Date() < graceEnd) return true
  }

  return false
}

File: apps/backend/src/yomi/services/entitlements.py
"""Entitlements: effective plan/role, renewal dates, billable access.

Port of apps/backend/src/entitlements.ts.
"""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import TypedDict

from yomi.shared.plans import PLANS, PlanConfig, get_plan
from yomi.shared.plans import feature_limit as plan_feature_limit

TRIAL_MS = 30 * 24 * 60 * 60 * 1000


class EntitlementUser(TypedDict, total=False):
    id: str | None
    email: str | None
    role: str | None
    plan: str | None
    subscription_status: str | None
    current_period_end: datetime | None
    trial_end_date: datetime | None
    created_at: datetime | None


def effective_role_for_user(user: EntitlementUser) -> str:
    return user.get("role") or "user"


def effective_plan_for_user(user: EntitlementUser) -> str:
    plan = user.get("plan") or "explore"
    return plan if plan in PLANS else "explore"


def get_plan_config(user: EntitlementUser) -> PlanConfig:
    return get_plan(effective_plan_for_user(user))


def request_limit_for_user(user: EntitlementUser) -> int | None:
    return plan_feature_limit(effective_plan_for_user(user), "chat")


def feature_limit_for_user(user: EntitlementUser, feature: str) -> int | None:
    return plan_feature_limit(effective_plan_for_user(user), feature)  # type: ignore[arg-type]


def _now() -> datetime:
    return datetime.now(UTC)


def _as_utc(value: datetime | None) -> datetime | None:
    return value


def credit_renewal(user: EntitlementUser) -> tuple[str, datetime | None]:
    """Returns (kind, at) where kind is "renewal" | "none"."""
    if effective_plan_for_user(user) == "explore":
        trial_end = user.get("trial_end_date")
        if trial_end is None and user.get("created_at"):
            trial_end = user["created_at"] + timedelta(milliseconds=TRIAL_MS)
        return ("renewal", trial_end) if trial_end else ("none", None)
    cpe = user.get("current_period_end")
    return ("renewal", cpe) if cpe else ("none", None)


def has_billable_plan_access(user: EntitlementUser) -> bool:
    """Explore: trialEndDate not reached. Paid: active/trialing, or past_due within 7-day grace."""
    plan = effective_plan_for_user(user)
    now = _now()

    if plan == "explore":
        trial_end = user.get("trial_end_date")
        if trial_end is None:
            return False
        end = trial_end.replace(tzinfo=UTC) if trial_end.tzinfo is None else trial_end
        return now < end

    status = user.get("subscription_status")
    if status in ("active", "trialing"):
        return True

    if status == "past_due":
        ref = user.get("current_period_end") or now
        ref = ref.replace(tzinfo=UTC) if ref.tzinfo is None else ref
        grace_end = ref + timedelta(days=7)
        if now < grace_end:
            return True

    return False